libraries/chain/controller.cpp

```diff
@@ -130,14 +134,40 @@ rpc::chain::submit_block_response controller_impl::submit_block( const rpc::chai
          LOG(info) << "Pushing block - Height: " << request.block.header.height
             << ", ID: " << request.block.id;
       }
+
       block_node = _state_db.create_writable_node( request.block.header.previous, request.block.id );
-      KOINOS_ASSERT( block_node, unknown_previous_block, "Unknown previous block" );
+
+      // If this is not the genesis case, we must ensure that the proposed block timestamp is greater
+      // than the parent block timestamp.
+      if ( block_node && !crypto::multihash_is_zero( request.block.header.previous ) )
+      {
+         apply_context parent_ctx;
+
+         parent_ctx.push_frame( stack_frame {
+            .call = crypto::hash( CRYPTO_RIPEMD160_ID, "submit_block"s ).digest,
+            .call_privilege = privilege::kernel_mode
+         } );
+
+         auto parent_node = _state_db.get_node( request.block.header.previous );
+         parent_ctx.set_state_node( parent_node );
+         time_lower_bound = system_call::get_head_block_time( parent_ctx ).t;
+      }
    }
 
    apply_context ctx;
 
    try
    {
+      // Genesis case, when the first block is submitted the previous must be the zero hash
+      if ( crypto::multihash_is_zero( request.block.header.previous ) )
```

Comment:
Why do we repeat the `multihash_is_zero` check down here when we care about the genesis case for checking time as well.

What about

``` c++
if ( crypto::multihash_is_zero( request.block.header.previous )
{
   KOINOS_ASSERT( request.block.header.height == 1, root_height_mismatch, "First block must have height of 1" );
}
else if ( block_node )
{
   // Time travel stuff here
}
```